File: tools/Cpp/t4-cpp-api/src/big_int.cpp
#include "t4decoder/big_int.hpp"

#include <algorithm>
#include <stdexcept>

namespace t4 {

void BigInt::trim() {
  while (!mag_.empty() && mag_.back() == 0) mag_.pop_back();
  if (mag_.empty()) neg_ = false;  // canonical zero is non-negative
}

BigInt BigInt::fromU64(std::uint64_t v) {
  BigInt r;
  while (v != 0) {
    r.mag_.push_back(static_cast<std::uint32_t>(v % kBase));
    v /= kBase;
  }
  return r;
}

BigInt::BigInt(long long v) {
  bool n = v < 0;
  // Take magnitude safely (handles LLONG_MIN without UB).
  unsigned long long u = n ? (~static_cast<unsigned long long>(v) + 1ull)
                           : static_cast<unsigned long long>(v);
  *this = fromU64(u);
  neg_ = (!mag_.empty()) && n;
}

BigInt BigInt::fromDecimalString(const std::string& s) {
  std::size_t i = 0;
  bool n = false;
  if (i < s.size() && (s[i] == '+' || s[i] == '-')) {
    n = (s[i] == '-');
    ++i;
  }
  if (i >= s.size())
    throw std::invalid_argument("BigInt::fromDecimalString: no digits");
  // Build by processing 9 decimal digits at a time from the most-significant
  // end via repeated mul-by-10^k + add. Simpler: accumulate digit by digit.
  BigInt r;
  const BigInt ten(10);
  for (; i < s.size(); ++i) {
    char c = s[i];
    if (c < '0' || c > '9')
      throw std::invalid_argument("BigInt::fromDecimalString: bad digit");
    r = r.mul(ten).add(BigInt(static_cast<long long>(c - '0')));
  }
  r.neg_ = (!r.mag_.empty()) && n;
  return r;
}

BigInt BigInt::powerOfTen(unsigned e) {
  // 10^e. Build directly in base-1e9 limbs: 10^(9*q) shifts q limbs, 10^r fills.
  BigInt r;
  unsigned q = e / 9;
  unsigned rem = e % 9;
  std::uint32_t lead = 1;
  for (unsigned k = 0; k < rem; ++k) lead *= 10u;
  r.mag_.assign(q, 0u);
  r.mag_.push_back(lead);
  r.trim();
  return r;
}

BigInt BigInt::abs() const {
  BigInt r = *this;
  r.neg_ = false;
  return r;
}

BigInt BigInt::negated() const {
  BigInt r = *this;
  if (!r.mag_.empty()) r.neg_ = !r.neg_;
  return r;
}

int BigInt::cmpMag(const BigInt& a, const BigInt& b) {
  if (a.mag_.size() != b.mag_.size())
    return a.mag_.size() < b.mag_.size() ? -1 : 1;
  for (std::size_t i = a.mag_.size(); i-- > 0;) {
    if (a.mag_[i] != b.mag_[i]) return a.mag_[i] < b.mag_[i] ? -1 : 1;
  }
  return 0;
}

int BigInt::cmp(const BigInt& a, const BigInt& b) {
  int sa = a.sign(), sb = b.sign();
  if (sa != sb) return sa < sb ? -1 : 1;
  if (sa == 0) return 0;
  int m = cmpMag(a, b);
  return sa > 0 ? m : -m;  // both negative => reverse magnitude order
}

BigInt BigInt::addMag(const BigInt& a, const BigInt& b) {
  BigInt r;
  std::uint64_t carry = 0;
  std::size_t n = std::max(a.mag_.size(), b.mag_.size());
  r.mag_.reserve(n + 1);
  for (std::size_t i = 0; i < n; ++i) {
    std::uint64_t s = carry;
    if (i < a.mag_.size()) s += a.mag_[i];
    if (i < b.mag_.size()) s += b.mag_[i];
    r.mag_.push_back(static_cast<std::uint32_t>(s % kBase));
    carry = s / kBase;
  }
  if (carry) r.mag_.push_back(static_cast<std::uint32_t>(carry));
  r.trim();
  return r;
}

// Requires |a| >= |b|.
BigInt BigInt::subMag(const BigInt& a, const BigInt& b) {
  BigInt r;
  std::int64_t borrow = 0;
  r.mag_.reserve(a.mag_.size());
  for (std::size_t i = 0; i < a.mag_.size(); ++i) {
    std::int64_t s = static_cast<std::int64_t>(a.mag_[i]) - borrow;
    if (i < b.mag_.size()) s -= b.mag_[i];
    if (s < 0) {
      s += kBase;
      borrow = 1;
    } else {
      borrow = 0;
    }
    r.mag_.push_back(static_cast<std::uint32_t>(s));
  }
  r.trim();
  return r;
}

BigInt BigInt::add(const BigInt& o) const {
  if (neg_ == o.neg_) {
    BigInt r = addMag(*this, o);
    r.neg_ = (!r.mag_.empty()) && neg_;
    return r;
  }
  // Differing signs => subtract smaller magnitude from larger.
  int m = cmpMag(*this, o);
  if (m == 0) return BigInt();  // cancel to zero
  if (m > 0) {
    BigInt r = subMag(*this, o);
    r.neg_ = (!r.mag_.empty()) && neg_;
    return r;
  }
  BigInt r = subMag(o, *this);
  r.neg_ = (!r.mag_.empty()) && o.neg_;
  return r;
}

BigInt BigInt::sub(const BigInt& o) const { return add(o.negated()); }

BigInt BigInt::mul(const BigInt& o) const {
  if (mag_.empty() || o.mag_.empty()) return BigInt();
  BigInt r;
  r.mag_.assign(mag_.size() + o.mag_.size(), 0u);
  for (std::size_t i = 0; i < mag_.size(); ++i) {
    std::uint64_t carry = 0;
    std::uint64_t ai = mag_[i];
    for (std::size_t j = 0; j < o.mag_.size(); ++j) {
      std::uint64_t cur = r.mag_[i + j] + ai * o.mag_[j] + carry;
      r.mag_[i + j] = static_cast<std::uint32_t>(cur % kBase);
      carry = cur / kBase;
    }
    std::size_t k = i + o.mag_.size();
    while (carry) {
      std::uint64_t cur = r.mag_[k] + carry;
      r.mag_[k] = static_cast<std::uint32_t>(cur % kBase);
      carry = cur / kBase;
      ++k;
    }
  }
  r.neg_ = neg_ != o.neg_;
  r.trim();
  return r;
}

BigInt BigInt::divModScalar(std::uint64_t d, std::uint64_t& rem) const {
  if (d == 0) throw std::invalid_argument("BigInt::divModScalar: divide by zero");
  // Guard the no-128-bit invariant: (d-1)*kBase + (kBase-1) must fit in uint64.
  // That holds for d up to ~1.8e10, far above any market denominator / 10^k<=9.
  if (d > (UINT64_MAX - (kBase - 1)) / kBase + 1)
    throw std::invalid_argument("BigInt::divModScalar: divisor too large");
  BigInt q;
  q.mag_.assign(mag_.size(), 0u);
  std::uint64_t carry = 0;
  for (std::size_t i = mag_.size(); i-- > 0;) {
    std::uint64_t cur = carry * kBase + mag_[i];
    q.mag_[i] = static_cast<std::uint32_t>(cur / d);
    carry = cur % d;
  }
  q.neg_ = neg_;
  q.trim();
  rem = carry;
  return q;
}

std::string BigInt::toString() const {
  if (mag_.empty()) return "0";
  std::string s;
  if (neg_) s.push_back('-');
  // Most-significant limb without padding, the rest zero-padded to 9 digits.
  s += std::to_string(mag_.back());
  for (std::size_t i = mag_.size() - 1; i-- > 0;) {
    std::string chunk = std::to_string(mag_[i]);
    s.append(9 - chunk.size(), '0');
    s += chunk;
  }
  return s;
}

}  // namespace t4

File: tools/Rust/t4-Rust-api/src/big_int.rs
//! Minimal arbitrary-precision signed integer (dependency-free).
//!
//! Port of `big_int.{hpp,cpp}`. Backs [`Decimal`](crate::Decimal) /
//! [`Price`](crate::Price). Only the subset the chart decoder needs is
//! implemented — notably **no general BigInt/BigInt long division** (the decode
//! path never needs it). Representation is sign-magnitude with base-1e9
//! little-endian limbs, making decimal rendering and power-of-ten scaling cheap.

use core::cmp::Ordering;
use core::fmt;

/// 1e9 per limb.
const KBASE: u32 = 1_000_000_000;
const KBASE64: u64 = KBASE as u64;

/// Sign-magnitude big integer with base-1e9 limbs.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct BigInt {
    neg: bool,
    /// Little-endian, each limb `< KBASE`; empty means zero.
    mag: Vec<u32>,
}

impl BigInt {
    /// The value zero.
    pub fn zero() -> Self {
        BigInt::default()
    }

    /// Construct from an unsigned 64-bit value.
    pub fn from_u64(mut v: u64) -> Self {
        let mut mag = Vec::new();
        while v != 0 {
            mag.push((v % KBASE64) as u32);
            v /= KBASE64;
        }
        BigInt { neg: false, mag }
    }

    /// Construct from a signed 64-bit value.
    pub fn from_i64(v: i64) -> Self {
        let n = v < 0;
        let mut r = BigInt::from_u64(v.unsigned_abs());
        r.neg = !r.mag.is_empty() && n;
        r
    }

    /// Parse an optional leading `+`/`-` then decimal digits.
    pub fn from_decimal_str(s: &str) -> Result<Self, String> {
        let bytes = s.as_bytes();
        let mut i = 0;
        let mut neg = false;
        if let Some(&c) = bytes.first() {
            if c == b'+' || c == b'-' {
                neg = c == b'-';
                i = 1;
            }
        }
        if i >= bytes.len() {
            return Err("BigInt::from_decimal_str: no digits".into());
        }
        let ten = BigInt::from_i64(10);
        let mut r = BigInt::zero();
        for &c in &bytes[i..] {
            if !c.is_ascii_digit() {
                return Err("BigInt::from_decimal_str: bad digit".into());
            }
            r = r.mul(&ten).add(&BigInt::from_i64((c - b'0') as i64));
        }
        r.neg = !r.mag.is_empty() && neg;
        Ok(r)
    }

    /// `10^e`.
    pub fn power_of_ten(e: u32) -> Self {
        let q = (e / 9) as usize;
        let rem = e % 9;
        let mut lead = 1u32;
        for _ in 0..rem {
            lead *= 10;
        }
        let mut mag = vec![0u32; q];
        mag.push(lead);
        let mut r = BigInt { neg: false, mag };
        r.trim();
        r
    }

    /// True when the value is zero.
    pub fn is_zero(&self) -> bool {
        self.mag.is_empty()
    }

    /// `-1`, `0`, or `+1`.
    pub fn sign(&self) -> i32 {
        if self.mag.is_empty() {
            0
        } else if self.neg {
            -1
        } else {
            1
        }
    }

    /// Absolute value.
    pub fn abs(&self) -> Self {
        BigInt {
            neg: false,
            mag: self.mag.clone(),
        }
    }

    /// Negation (zero stays non-negative).
    pub fn negated(&self) -> Self {
        BigInt {
            neg: !self.mag.is_empty() && !self.neg,
            mag: self.mag.clone(),
        }
    }

    /// Signed three-way compare.
    pub fn compare(a: &BigInt, b: &BigInt) -> Ordering {
        let (sa, sb) = (a.sign(), b.sign());
        if sa != sb {
            return sa.cmp(&sb);
        }
        if sa == 0 {
            return Ordering::Equal;
        }
        let m = BigInt::cmp_mag(a, b);
        if sa > 0 {
            m
        } else {
            m.reverse()
        }
    }

    /// Sum.
    pub fn add(&self, o: &BigInt) -> Self {
        if self.neg == o.neg {
            let mut r = BigInt::add_mag(self, o);
            r.neg = !r.mag.is_empty() && self.neg;
            return r;
        }
        match BigInt::cmp_mag(self, o) {
            Ordering::Equal => BigInt::zero(),
            Ordering::Greater => {
                let mut r = BigInt::sub_mag(self, o);
                r.neg = !r.mag.is_empty() && self.neg;
                r
            }
            Ordering::Less => {
                let mut r = BigInt::sub_mag(o, self);
                r.neg = !r.mag.is_empty() && o.neg;
                r
            }
        }
    }

    /// Difference.
    pub fn sub(&self, o: &BigInt) -> Self {
        self.add(&o.negated())
    }

    /// Product.
    pub fn mul(&self, o: &BigInt) -> Self {
        if self.mag.is_empty() || o.mag.is_empty() {
            return BigInt::zero();
        }
        let mut mag = vec![0u32; self.mag.len() + o.mag.len()];
        for i in 0..self.mag.len() {
            let mut carry: u64 = 0;
            let ai = self.mag[i] as u64;
            for j in 0..o.mag.len() {
                let cur = mag[i + j] as u64 + ai * o.mag[j] as u64 + carry;
                mag[i + j] = (cur % KBASE64) as u32;
                carry = cur / KBASE64;
            }
            let mut k = i + o.mag.len();
            while carry != 0 {
                let cur = mag[k] as u64 + carry;
                mag[k] = (cur % KBASE64) as u32;
                carry = cur / KBASE64;
                k += 1;
            }
        }
        let mut r = BigInt {
            neg: self.neg != o.neg,
            mag,
        };
        r.trim();
        r
    }

    /// Divide `|self|` by a positive scalar `d`, returning `(quotient, remainder)`.
    ///
    /// The quotient carries `self`'s sign; the remainder is always `>= 0`. `d`
    /// must be small enough that `(d-1)*KBASE + (KBASE-1)` fits in `u64` (holds
    /// for `d` up to ~1.8e10, far above any market denominator or `10^k, k<=9`).
    pub fn div_mod_scalar(&self, d: u64) -> (BigInt, u64) {
        assert!(d != 0, "BigInt::div_mod_scalar: divide by zero");
        assert!(
            d <= (u64::MAX - (KBASE64 - 1)) / KBASE64 + 1,
            "BigInt::div_mod_scalar: divisor too large"
        );
        let mut mag = vec![0u32; self.mag.len()];
        let mut carry: u64 = 0;
        for i in (0..self.mag.len()).rev() {
            let cur = carry * KBASE64 + self.mag[i] as u64;
            mag[i] = (cur / d) as u32;
            carry = cur % d;
        }
        let mut q = BigInt { neg: self.neg, mag };
        q.trim();
        (q, carry)
    }

    // --- private helpers ---------------------------------------------------

    fn trim(&mut self) {
        while matches!(self.mag.last(), Some(&0)) {
            self.mag.pop();
        }
        if self.mag.is_empty() {
            self.neg = false;
        }
    }

    fn cmp_mag(a: &BigInt, b: &BigInt) -> Ordering {
        if a.mag.len() != b.mag.len() {
            return a.mag.len().cmp(&b.mag.len());
        }
        for i in (0..a.mag.len()).rev() {
            if a.mag[i] != b.mag[i] {
                return a.mag[i].cmp(&b.mag[i]);
            }
        }
        Ordering::Equal
    }

    fn add_mag(a: &BigInt, b: &BigInt) -> BigInt {
        let n = a.mag.len().max(b.mag.len());
        let mut mag = Vec::with_capacity(n + 1);
        let mut carry: u64 = 0;
        for i in 0..n {
            let mut s = carry;
            if i < a.mag.len() {
                s += a.mag[i] as u64;
            }
            if i < b.mag.len() {
                s += b.mag[i] as u64;
            }
            mag.push((s % KBASE64) as u32);
            carry = s / KBASE64;
        }
        if carry != 0 {
            mag.push(carry as u32);
        }
        let mut r = BigInt { neg: false, mag };
        r.trim();
        r
    }

    /// Requires `|a| >= |b|`.
    fn sub_mag(a: &BigInt, b: &BigInt) -> BigInt {
        let mut mag = Vec::with_capacity(a.mag.len());
        let mut borrow: i64 = 0;
        for i in 0..a.mag.len() {
            let mut s = a.mag[i] as i64 - borrow;
            if i < b.mag.len() {
                s -= b.mag[i] as i64;
            }
            if s < 0 {
                s += KBASE64 as i64;
                borrow = 1;
            } else {
                borrow = 0;
            }
            mag.push(s as u32);
        }
        let mut r = BigInt { neg: false, mag };
        r.trim();
        r
    }
}

impl From<i64> for BigInt {
    fn from(v: i64) -> Self {
        BigInt::from_i64(v)
    }
}

impl fmt::Display for BigInt {
    /// Signed decimal, no leading zeros.
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.mag.is_empty() {
            return f.write_str("0");
        }
        if self.neg {
            f.write_str("-")?;
        }
        // Most-significant limb unpadded, the rest zero-padded to 9 digits.
        write!(f, "{}", self.mag[self.mag.len() - 1])?;
        for i in (0..self.mag.len() - 1).rev() {
            write!(f, "{:09}", self.mag[i])?;
        }
        Ok(())
    }
}
